src/lib/shmem/src/allocator.rs

```diff
@@ -92,6 +80,26 @@ where
     extern "C" fn _validate_stable_layout(_: T) {}
 }
 
+impl<'origin, T> Drop for ShMemBlock<'origin, T>
+where
+    T: Sync + VirtualAddressSpaceIndependent,
+{
+    fn drop(&mut self) {
+        unsafe { std::ptr::drop_in_place(self.internal.p as *mut T) };
```

Comment:
Also add a SAFETY comment to `ShMemBlock::new()` stating that you must not create two `ShMemBlock` that point to the same shared memory?

Reply:
I think it would be clearer if it was split, since then you have some ideas about what things you should uphold based on the type. For example if you have a `ReferencedShMemBlock`, you know you should drop it before any `OwnedShMemBlock`. But definitely doesn't need to be changed for this PR.

Reply:
SG; I'll do a follow-up